### PR TITLE
Measure service health based on weights reported by haproxy

### PR DESCRIPTION
**Pull request by mrak**
This adds a new `haproxy_weight` (`Metric::HaproxyWeight`) measure for a service that will interrogate a haproxy stats socket for backend weighting information and report their average.

```ruby
# /etc/litmus.d/test.conf
service "test" do |s|
  s.depends(Dependency::HaproxyBackends,
            "/var/run/test-haproxy-stats.sock",
            "test-backend")

  s.measure_health_with_args(Metric::HaproxyWeight, 
                             "/var/run/test-haproxy-stats.sock",
                             "test-backend")
end
```

Assuming that you have three `server` lines in you haproxy configuration with weights `20`, `80`, and `80`, this measure would report `(20+80+80)/3 = 60`. 

This allows haproxy weighting to be reflected as the reported health of a litmus service.

### CODE REVIEW

File: lib/litmus_paper/dependency/haproxy_backends.rb [hunk 1]
```diff
@@ -28,13 +28,31 @@ def available?
...
       end
 
+      def average_weight
+        stats = _parse_stats(_fetch_stats)
+        backend = _find_backend(stats, @cluster)
+
+        total_weight = backend['weight'].to_i
+        total_servers = backend['act'].to_i
+
+        return total_servers == 0 ? 0 : total_weight / total_servers
```
[the remark this answers is not in the document]
Comment (mrak) [reply lib/litmus_paper/dependency/haproxy_backends.rb]: Our first use case would be setting authy pod weights in production to `1`, each. This metric would report `1` as the litmus health on the kube-lbs. If it were a sum, if we had 10 authy pods the number would report around `10`.

We are planning to use this as a way to specify a trickle of `1%` traffic to authy pods in production without having to modify production load balancers with hardcoded trickles/weights.

[the remark this answers is not in the document]
Comment (lollipopman) [reply lib/litmus_paper/dependency/haproxy_backends.rb]: That makes sense. Most of the other metrics accept a weight parameter, so the user can scale the weight to their preferred range, for our use cases that is normally 0-100. Should we perhaps add a weight parameter and divide the average weight by the maximum HAProxy weight? e.g. for a backend with two servers of weight 30. `weight * 60 / (2 * 256)`

[the remark this answers is not in the document]
Comment (bjhaid) [reply lib/litmus_paper/dependency/haproxy_backends.rb]: where servers == k8s pods in this case

**Comment by duelinmarkers** [comment]
Do I understand correctly that this PR treats haproxy server weights as a unit that can be compared across instances, when haproxy's existing semantic for weights is that they're only meaningful in comparison to other weights on servers in the same backend? For example, haproxy sees no difference between a backed with servers weighted 1, 2, 2 and one with servers weighted 5, 10, 10 or 128, 256, 256, but this metric would report very different litmus health for each of those.

I think this use of weights would generally be surprising to people familiar with haproxy server weights and should probably carry a stranger name than `Metric::HaproxyWeight` to try to communicate the new meaning it introduces to an existing haproxy configuration parameter.

**Comment by duelinmarkers** [comment]
My concern stands: this is a surprising way to use haproxy weight, since it adds a layer of meaning different from what it means to haproxy. I'm not saying it shouldn't be possible, but the intuitive way to derive server health from weights would be `sum(weights of UP servers in backend) / sum(weights of all servers in backend)`.

Does this also require weights to be <=100 to result in a reasonable health (whereas the max in haproxy is 256) or is litmus health not limited to 0-100?

**Comment by bjhaid** [comment]
> sum(weights of UP servers in backend) / sum(weights of all servers in backend)

Also this does not provide much value for what we want to do, we'll explore alternate approaches to solving our problems thanks for the feedback

**Comment by duelinmarkers** [comment]
>> sum(weights of UP servers in backend) / sum(weights of all servers in backend)
>
> This is always 1, you can verify haproxy does not include weight of servers not UP in the summed weight that this PR is using.

You misunderstand me. I'm saying the health should be the proportion of up weight compared to total configured weight for the backend. I'm not referring to any specific sum haproxy reports.

**Comment by bjhaid** [comment]
> You misunderstand me. I'm saying the health should be the proportion of up weight compared to total configured weight for the backend. I'm not referring to any specific sum haproxy reports.

The maths does not match what we intend doing, we want to use this to force trickling of traffic and ensure whatever arbitrary weight we program in haproxy propagates all the way to ipvs. So what you are suggesting is a completely different feature set from what we want

**Pull request closed without merging**